### PR TITLE
feat: store blocks by multihash instead of CID

### PR DESCRIPTION
**Pull request by achingbrain**
Updates the `ipfs-repo` dep to a version that stores blocks by multihash instead of CID to support CIDv1 and CIDv0 access to the same block.

New features:

- Adds a `--multihash` argument to the cli command `ipfs refs local` which prints the base32 encoded multihash of each block

BREAKING CHANGES:

- `ipfs.refs.local` now returns a v1 CID with the raw codec for every block and not the original CID by which it was added to the blockstore

### CODE REVIEW

**Comment by achingbrain** [comment]
Supersedes #2522

**Comment by achingbrain** [comment]
@aschmahmann could I get a 👍 or 👎  on the cli change please?  It would be a shame to ship this and then have go-ipfs implement something different later down the line.

The change is:

* `ipfs refs local` used to print out CIDs of blocks in your repo
* Now it prints v1 raw CIDs instead
* There's a new `--multihash` argument to `ipfs refs local` which prints base32 encoded multihashes instead of the v1 raw CIDs

**Comment by aschmahmann** [comment]
@achingbrain I'm having trouble finding which issue describes what we agreed on here. This change seems reasonable to me, but mostly because I'm not sure what people are using `ipfs refs local` for... I recall we use `ipfs refs {cid}` with the preload nodes but that code path is left unchanged.

Any pointers?

**Comment by achingbrain** [comment]
The issue & decision is here: https://github.com/ipfs/specs/issues/242#issuecomment-627533056

Also relevant: https://github.com/ipfs/go-ipfs/issues/6815

**Comment by aschmahmann** [comment]
Neither of those issues seem to explicitly reference the changes to `ipfs refs local`. Although I guess it's probably implied by those issues since there may be orphaned (e.g. unreferenced by a pin) blocks that we just can't figure out the CIDs for.

**Comment by achingbrain** [comment]
> I guess it's probably implied by those issues since there may be orphaned (e.g. unreferenced by a pin) blocks that we just can't figure out the CIDs for

Exactly, it's more of a practical implication of dropping the original CID a block was stored by.

File: packages/ipfs/src/cli/commands/refs-local.js [hunk 1]
```diff
@@ -1,27 +1,40 @@
...
       timeout
     })) {
-      if (ref.err) {
-        print(ref.err, true, true)
+      if (err) {
+        print(err, true, true)
       } else {
-        print(ref.ref)
+        if (multihash) {
+          print(multibase.encode('base32', Buffer.from(ref)).toString().substring(1).toUpperCase())
```
Comment (hugomrdias) [inline packages/ipfs/src/cli/commands/refs-local.js:34]:
```suggestion
          print(multibase.encoding('base32upper').encode(Buffer.from(ref)))
```
with multibase v1 we can just do this or

```js
multibase.names['base32upper'].encode(Buffer.from(ref))
```

File: packages/ipfs/test/cli/refs-local.js [hunk 2]
```diff
@@ -37,6 +39,23 @@ describe('refs local', () => {
...
+    ipfs.refs.local.withArgs(defaultOptions).returns([{
+      ref
+    }, {
+      err
+    }])
+
+    const out = await cli('refs local --multihash', { ipfs })
+    const lines = out.split('\n')
+
+    expect(lines.includes(multibase.encode('base32', Buffer.from(ref)).toString().substring(1).toUpperCase())).to.be.true()
```
Comment (hugomrdias) [inline packages/ipfs/test/cli/refs-local.js:55]:
```suggestion
    expect(lines.includes(multibase.encoding('base32upper').encode(Buffer.from(ref))).to.be.true()
```

**Comment by aschmahmann** [comment]
Seems reasonable 👍 from me then. If @hsanjuan or @Stebalien have a second to confirm that'd be great, since they've been more involved with this than I have. However, if they're busy then no complaints from me

File: packages/ipfs/src/cli/commands/refs-local.js (hunk 1 above)
Comment (achingbrain) [reply packages/ipfs/src/cli/commands/refs-local.js:34]: Neato, I didn't realise.

**Comment by Stebalien** [comment]
@achingbrain that's the planned change for go-ipfs as well.

**Comment by oed** [comment]
So does this change mean that all data already stored in an ipfs node will not be retrievable locally when one updates to v0.48? 
Is there a data migration strategy that can be used here?

**Comment by achingbrain** [comment]
No, there should be no change. A [blockstore migration](https://github.com/ipfs/js-ipfs-repo-migrations/pull/4) takes place when you first start a node running 0.48.0 on a non-migrated repo to convert all the keys so what is currently retrievable should still be retrievable after the upgrade.

**Pull request merged**